### DISCUSSION TITLE
How to handle custom signin page with next-auth

### DISCUSSION THREAD

**Discussion by mkko120** [nextauthjs/next-auth#8199] [Help]
### Question 💬

I am currently building an app with next-js' app router and i want to have a custom login page because the default one is not centered. The problem is that after signing in (somehow) i am still redirected to the login page. How should i configure the router to handle the authorization properly? There is not much documentation about it so thats why i'm asking.

### How to reproduce ☕️

/api/[...nextauth]/route.ts
```jsx
import NextAuth from "next-auth"
import DiscordProvider from "next-auth/providers/discord";

const handler = NextAuth({
    // Configure one or more authentication providers
    providers: [
        DiscordProvider({
            name: "Discord",
            clientId: process.env.DISCORD_CLIENT_ID || "",
            clientSecret: process.env.DISCORD_CLIENT_SECRET || "",
        })
    ],
    session: {
        strategy: "jwt",
    },
    pages: {
        signIn: "/auth/signin",
    }
})

export { handler as GET, handler as POST }
```
middleware.ts
```ts
export { default } from "next-auth/middleware"

export const config = {
    matcher: [
        "/panel",
        "/rekrutacja",
        "/zwolnienia",
    ],
    pages: {
        signIn: "/auth/signin",
    }
}
```
/auth/signin/page.tsx
```tsx
"use client";

import {Button, Center, Container, Group} from "@mantine/core";
import {signIn} from "next-auth/react";
import {IconBrandDiscord, IconBrandDiscordFilled} from "@tabler/icons-react";

export default function LoginPage() {
    return (
        <Container className="h-screen">
            <Center className="h-full flex-col">
                <h1>Witaj, <b>nieznajomy!</b></h1>
                <span>Zaloguj się aby przejść dalej</span>

                <Button variant="outline" onClick={() => signIn("discord")} className="mt-5" color="indigo">
                    <Group>
                        <IconBrandDiscordFilled />
                        <span>Zaloguj się przez Discord</span>
                    </Group>
                </Button>
            </Center>

        </Container>
    )
}
```

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by JinMABrett** [comment]
I apologize in advance if this is not what you are looking for, but since it's been 3 days and no response, i just thought id pass this along that I was using earlier. Sorry again if this is not quite what you were asking for

https://www.youtube.com/watch?v=hADeo48SATU&

**Comment by mkko120** [comment]
I was initially looking at that video, but unfortunately it uses the pages router to handle next-auth :(

**Comment by xCoBaLTz** [comment]
I am having the exact same error, custom login pages using the app router seem to redirect back to the login page. Also was referring to that video.

**Comment by jules-sommer** [comment]
Hey guys, just wanted to chime in here with what I think is a reasonable solution to this behaviour and is what I implement in my production app; from my brief read through of this issue I understand you're having the issue of being redirected to the sign-in page after logging in via triggering a flow for your respective OAuth provider with "next-auth/react" { signIn } function. You then login with the provider and end up back on on your custom sign in page.

What I find is best practice, since you don't want users being able to login or interact with the login form when they're already signed in, is to just wrap your login page in a server component, if it's not one already, and use getServerSession and some simple logic to either render the login form or redirect using "next/navigation"s redirect() function. Something like this:


```
import { redirect } from 'next/navigation'
import { authOptions } from 'file exporting NextAuth({ configObj })'
import { getServerSession } from 'next-auth'

const LoginPage = async () => {
    const session = await getServerSession(authOptions)
    // if the user is not logged in, show the login page, otherwise redirect to the account page
    if (!session) {
        return (
            <Login />
        );
    } else {
        redirect('/account', 'replace')
    }
}

export default LoginPage
```

That way, you're quickly redirected without the login form being rendered to the appropriate page, and have the benefit of authed users not being able to navigate back to the login page manually. I suppose this could also be replicated client side with the useSession hook, but that would likely cause more wait times for users as the useSession hook has to make a round trip to the server to verify that token.

Let me know if this helped or if I missed the mark on what exactly your issue was.

**Comment by ThangHuuVu** [comment]
It seems that you are setting the `pages` option in the middleware incorrectly, it should be: 
```js
import { withAuth } from "next-auth/middleware"

export default withAuth(
  {
    pages: {
      signIn: "/auth/signin",
    },
  }
)
export const config = {
    matcher: [
        "/panel",
        "/rekrutacja",
        "/zwolnienia",
    ],
}
```
